### ISSUE TITLE
Demonstrating Proof Of Possession (DPoP)

### ISSUE THREAD

**Issue by dbfr3qs** [authts/oidc-client-ts#1360]
Recently, the IETF released [RFC9449](https://datatracker.ietf.org/doc/html/rfc9449), a proposed internet standard that describes a method for constraining access tokens to devices through the use of public key cryptography.

The idea is for the client to generate a cryptographic key pair, and use this to sign requests to the authorisation server for access tokens/refresh tokens, and then on subsequent requests to resources with said access tokens/refresh tokens. These signatures are referred to as "proofs" and are supplied in an additional request header. The spec also allows for optionally constraining authorisation codes, so that the entire end to end process is constrained and there's no chance anything can be used outside of the client.

In the frontend, this could be particularly useful because there's no secure method to store tokens, opening up vectors for token exfiltration, for example through XSS or other means (e.g. supply chain, CDN cache pollution). DPoP may provide a way to eliminate certain attack vectors which seek to exfiltrate tokens out of the browser to be used by separate systems owned by bad actors, though it would not protect against attacks where [a compromised browser is used directly by a bad actor](https://www.ietf.org/archive/id/draft-ietf-oauth-browser-based-apps-15.html#section-6.3.3.2.2).

To prevent leaking/exfiltration of the key material used to generate the DPoP proofs in the browser, we can make use of the browser's SubtleCrypto api to generate cryptographic keys that are not extractable. From there, [we can store the CryptoKeyPair object using indexedDB](https://crypto.stackexchange.com/a/52488/113803), and use that object to generate DPoP proofs. This provides a way to securely store key material without the possibility of exfiltration, though I am not sure it would prevent a motivated bad actor who had gain full control of a user's machine or was [proxying requests through a compromised browser](https://www.ietf.org/archive/id/draft-ietf-oauth-browser-based-apps-15.html#name-proxying-requests-via-the-u).

As cookie based authentication becomes more and more restricted by browsers, and refresh tokens become more standard to keep sessions alive, it is my opinion that constraining tokens will become essential, particularly in pure frontend clients.

Is this something that is worth considering adding to the library?

I've provided a [PR](https://github.com/authts/oidc-client-ts/pull/1361) with an implementation that covers most of the spec, and uses the native SubtleCrypto api and indexedDB as described above. I have left out a few areas that I don't think are relevant to the use cases oidc-client-ts addresses. If this is something worth considering, please take a look.

**Comment by dbfr3qs** [comment]
Cool - I realise it's quite a lot to take in so keen to collaborate and contribute further if you think it's worthwhile functionality.

**Comment by uvdsl** [comment]
I'd like second DPoP support in this library. Would be really useful.

**Comment by Thomathan** [comment]
Hi, I noticed there is a merged PR for adding DPoP and is targeted for 3.1.0. Did that work satisfy this ticket and allow us to implement dpop into our auth flows when it releases?

**Comment by dbfr3qs** [comment]
Yes, it is currently in the release candidate build if you would like to try it out.

**Comment by indiealexh** [comment]
With DPOP enabled, it doesn't seem to send the DPOP header with the token request resulting in a Bad Request error from OIDC Server.

Are there instructions missing? Or is this a bug?
